### PR TITLE
Micropilot

### PR DESCRIPTION
**Pull request by monicachew**
This is just some placeholder code, but I could use some testing advice. Also please review the changes to the README to see if setup directions make sense.

### CODE REVIEW

**Comment by monicachew** [comment]
For testing, I'd like to somehow trigger the upload and check that the events emitted by the tests are what we expect. I guess I could expose forceUpload for testing purposes, then yield to the promise returned by monitor.upload. Is there a way to expose forceUpload only to the test?

File: package.json [hunk 1]
```diff
@@ -5,5 +5,21 @@
...
+    {
+      "type": "bool",
+      "title": "micropilotlog",
+      "name": "micropilotlog",
+      "value": true
+    }
+  ],
+  "volo": {
+    "dependencies": {
+      "packages/micropilot": "github:gregglind/micropilot/v0.8"
```
Comment (mozkeeler) [inline package.json:28]: I was expecting this to automatically fetch micropilot for me. Too much to ask, or do I not have volo set up correctly?

**Comment by mozkeeler** [comment]
I don't know of a way to expose something only to the tests. There are already some testing-only custom events that are emitted that in theory regular code could listen for, so I don't think it's that bad to expose a testing-only thing to non-tests. Maybe we should add annotations? Also, it would be cool to have a way of replacing the micropilot server with localhost:4444 and registering a handler that will check that the data it receives is good (the latter part should be pretty easy - see https://developer.mozilla.org/en-US/docs/Httpd.js/HTTP_server_for_unit_tests ).

File: package.json (hunk 1 above)
Comment (monicachew) [inline package.json:28]: I think, but am not sure, that this means that if we use volo to install blushproof itself then volo will fetch packages/micropilot, e.g.

volo create myaddon micropilot-template

However, I'm not sure how/if we can take advantage of that in the development cycle.

@gregglind, any clues?

**Comment by monicachew** [comment]
So, if simulate = true in the upload then we get the data back in the response, which I think is good enough for testing we are recording what we expect. I think we should differentiate between
1. Testing that micropilot upload works correctly (https://github.com/gregglind/micropilot/blob/master/test/test-micropilot-upload.js)
2. Testing that we record what we expect

Otherwise unit tests devolve into integration tests.

File: package.json (hunk 1 above)
Comment (monicachew) [inline package.json:28]: This doesn't seem to be supported: https://github.com/volojs/volo/issues/149

Seems like the intended workflow is to manually call volo add -f when you want to update the dependency.

Comment (monicachew) [inline package.json:28]: Hmm, figure out workflow by reading bugs: https://github.com/volojs/volo/issues/22

@jrburke, is the expected workflow to contribute to a git-managed codebase with volo dependencies to

git clone myproject
cd myproject
volo add

?

**Comment by monicachew** [comment]
I ended up doing a massive rewrite of the test code to use promises instead of continuation callbacks. It is shorter and reads more synchronously, so I think it's easier to understand -- but I am sorry it ended up being so large.

This is more or less ready for review.

File: test/test-main.js [hunk 2]
```diff
@@ -1,114 +1,101 @@
...
+      for (let i = 0; i < events.length; ++i) {
+        // Ignore the timestamp field
+        gAssert.equal(events[i].event, gEvents[i]);
+        // Micropilot sticks an eventstoreid field on every record
+        gAssert.equal(events[i].eventstoreid, i + 1);
+      }
+      deferred.resolve(true);
+      return deferred.promise;
+    }).
+    // This clear is not working.
```
Comment (mozkeeler) [inline test/test-main.js:45]: This is something I have a question about that we should discuss during the meeting today.

File: test/test-main.js [hunk 3]
```diff
@@ -116,43 +103,76 @@ function expectOrNotPanelCommon(aURL, aContinuation, aExpectPanel, aDoNav) {
...
-      if (aSuccess) {
-        let panel = aEvent.detail;
-        expectNoConsentPanelNoNav("http://localhost:4444/", runNextTest);
-        panel.postMessage("continue");
-      } else {
-        runNextTest();
-      }
+function maybePushBlushButton(win, aForget) {
+  let deferred = defer();
+  bpUI.blushButton.panel.show();
```
Comment (mozkeeler) [inline test/test-main.js]: Shouldn't this be after we add the event listeners at the bottom of this function?

File: test/test-main.js [hunk 4]
```diff
@@ -160,7 +180,10 @@ function testExpectConsentPanel() {
  * panel, because we whitelisted it in the previous test.
  */
 function testExpectNoConsentPanelWhitelisted() {
-  expectNoConsentPanel("http://localhost:4444/", runNextTest);
+  console.log("testExpectConsentPanelThenWhitelist");
```
Comment (mozkeeler) [inline test/test-main.js]: testExpectNoConsentPanelThenWhitelist, right? Also, should we ensure all of these console.log messages match the corresponding test names?

File: test/test-main.js [hunk 5]
```diff
@@ -170,187 +193,141 @@ function testExpectNoConsentPanelWhitelisted() {
 function testExpectNoConsentPanelNotOnBlushlist() {
   let key = bpUtil.getKeyForHost("localhost");
   delete ss.storage.blushlist.map[key];
-  // we have to clear these together to keep things consistent
+  // We have to clear these together to keep things consistent.
   delete ss.storage.whitelistedDomains[key];
   delete ss.storage.whitelistedCategories["testing"];
-  gAssertObject.equal(bpCategorizer.getCategoryForHost("localhost"),
+  gAssert.equal(bpCategorizer.getCategoryForHost("localhost"),
                       null,
```
Comment (mozkeeler) [inline test/test-main.js]: Indentation

File: test/test-main.js [hunk 6]
```diff
@@ -170,187 +193,141 @@ function testExpectNoConsentPanelWhitelisted() {
...
+    then(function(response) {
+      response.event.detail.hide();
+      return testMonitor(); });
 }
 
-// In case the function name isn't clear: this test checks that we properly
-// remove a domain from the blushlist if the user used the "Blush This!"
-// button on it.
+// Tests that we properly remove a domain from the blushlist if the user used
+// the "Blush This!" button on it.
```
Comment (mozkeeler) [inline test/test-main.js]: I guess this comment was never clear - I meant "if the user had previously used the 'Blush This!' button on it and then subsequently whitelisted it."

Also, we should probably be consistent in commenting styles (my bad) - most of these explain-the-function comments use the /\* */ style.

File: package.json [hunk 7]
```diff
@@ -5,5 +5,27 @@
...
+      "name": "enable_reporting",
+      "value": false
+    }
+  ],
+  "volo": {
+    "dependencies": {
+      "packages/micropilot": "github:gregglind/micropilot/v0.8"
+    }
+  }
+}
```
Comment (mozkeeler) [inline package.json]: So, there's no newline at the end of this file. Do we care?

File: test/test-main.js (hunk 3 above)
Comment (monicachew) [inline test/test-main.js]: Yes! Will move.

File: test/test-main.js (hunk 6 above)
Comment (monicachew) [inline test/test-main.js]: Good point, I'll fix this.

File: lib/monitor.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+const PULSE_INTERVAL = 24 * 60 * 60 * 1000;
+
+let monitor = micropilot.Micropilot("blushproof").start();
+/**
+ * A helper function to record time-stamped events. Times are modulo the
+ * most recent hour.
+ * @param eventName {string} The name of the event.
+ * @return promise
+ */
+//micropilot.Micropilot.prototype.recordEvent = function recordEvent(eventName) {
```
Comment (monicachew) [inline lib/monitor.js:20]: This didn't work

File: package.json (hunk 7 above)
Comment (monicachew) [inline package.json]: I got a trailing whitespace error when I tried to commit! Seems a little strict...

**Comment by monicachew** [comment]
Fixed everything (I think) besides the prototype issue, which I still haven't figured out. OK to merge?

**Pull request merged**